compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/generators/DependencyGraphFirGenerator.kt

```diff
@@ -403,9 +421,23 @@ internal class DependencyGraphFirGenerator(session: FirSession) :
                 name = parameter.name,
                 key = Keys.RegularParameter,
                 typeProvider = {
-                  parameter.resolvedReturnType.withArguments(
-                    it.mapToArray(FirTypeParameterRef::toConeType)
-                  )
+                  if (parameter.resolvedReturnType is ConeTypeParameterType) {
+                    val creator =
+                      graphObject(context.owner.requireContainingClassSymbol())
+                        ?.findCreator(
+                          session,
+                          "generateConstructors for ${context.owner.classId}",
+                          ::log,
+                        )
+                    parameter.resolveReturnTypeFrom(
+                      typeOwner = creator?.classSymbol!!,
+                      session = session,
+                    )
+                  } else {
+                    parameter.resolvedReturnType.withArguments(
+                      it.mapToArray(FirTypeParameterRef::toConeType)
+                    )
+                  }
```

Comment:
This part can be improved probably, but not sure what's the ideal way to do it. The problem is we need `creator` to resolve the actual generic type, I tried to make it as lazy as possible here